allauth_2fa/middleware.py

```diff
@@ -22,3 +27,83 @@ def process_request(self, request):
                 del request.session['allauth_2fa_user_id']
             except KeyError:
                 pass
+
+
+class BaseRequire2FAMiddleware(MiddlewareMixin):
+    """
+    Ensure that particular users have two-factor authentication enabled before
+    they have access to the rest of the app.
+
+    If they don't have 2FA enabled, they will be redirected to the 2FA
+    enrollment page and not be allowed to access other pages.
+    """
+
+    # List of URL names that the user should still be allowed to access.
+    allowed_pages = [
+        # They should still be able to log out or change password.
+        'account_change_password',
+        'account_logout',
+        'account_reset_password',
+
+        # URLs required to set up two-factor
+        'two-factor-setup',
+        'two-factor-qr-code',
+    ]
+    # The message to the user if they don't have 2FA enabled and must enable it.
+    require_2fa_message = "You must enable two-factor authentication before doing anything else."
+
+    def on_require_2fa(self, request):
+        """
+        If the current request requires 2FA and the user does not have it
+        enabled, this is executed. The result of this is returned from the
+        middleware.
+        """
+        # See allauth.account.adapter.DefaultAccountAdapter.add_message.
+        if 'django.contrib.messages' in settings.INSTALLED_APPS:
+            # If there is already a pending message related to two-factor (likely
+            # created by a redirect view), simply update the message text.
+            storage = messages.get_messages(request)
+            tag = '2fa_required'
+            for m in storage:
+                if m.extra_tags == tag:
+                    m.message = self.require_2fa_message
+                    break
+            # Otherwise, create a new message.
+            else:
+                messages.error(request, self.require_2fa_message, extra_tags=tag)
+            # Mark the storage as not processed so they'll be shown to the user.
+            storage.used = False
+
+        # Redirect user to two-factor setup page.
+        return redirect('two-factor-setup')
+
+    def require_2fa(self, request):
+        """
+        Check if this request is required to have 2FA before accessing the app.
+
+        This should return True if this request requires 2FA. (Note that the user was already)
+
+        You can access anything on the request, but generally request.user will
+        be most interesting here.
+        """
+        raise NotImplemented('You must implement test_request.')
```

Comment:
Should this be `You must implement require_2fa.`?

Reply:
Yes! Thanks. I had renamed this method at some point.